src/main/kotlin/skytils/skytilsmod/features/impl/mining/MiningFeatures.kt

```diff
@@ -195,29 +194,29 @@ class MiningFeatures {
-                val coords = parts[1].split(",")
-                if (locationsToInternal.containsKey(location)) {
-                    val key = locationsToInternal[location]
-
-                    /**
-                     * Sends the waypoints message except it suggests which one should be used based on
-                     * the name contained in the message and converts it to the internally used names for the waypoints.
-                     */
-                    UChat.chat(UTextComponent("§3Skytils > §eFound coordinates in a chat message, click a button to set a waypoint.\n")
-                        .appendSibling(UTextComponent("§f[${location}] ")
-                            .setClick(ClickEvent.Action.RUN_COMMAND, "/skytilshollowwaypoint set $key ${coords[0]} ${coords[1]} ${coords[2]}")
-                            .setHover(HoverEvent.Action.SHOW_TEXT, UTextComponent("§eset waypoint for $location"))
-                        )
-                        .appendSibling(UTextComponent("§e[Custom]")
-                            .setClick(ClickEvent.Action.SUGGEST_COMMAND, "/skytilshollowwaypoint set name ${coords[0]} ${coords[1]} ${coords[2]}")
-                            .setHover(HoverEvent.Action.SHOW_TEXT, UTextComponent("§eset custom waypoint")
-                            )
-                        )
-                    )
+            val cleanedUnformatted = unformatted.removeExceptPattern(SBE_DSM_PATTERN)
+            val matcher = SBE_DSM_PATTERN.matcher(cleanedUnformatted)
+            if (matcher.matches()) {
+                val stringLocation = matcher.group("stringLocation")
+                val x = matcher.group("x")
+                val y = matcher.group("y")
+                val z = matcher.group("z")
+                val loc = CrystalHollowsMap.Locations.values().find { it.cleanName == stringLocation }
+                if (loc != null) {
+                    if (!loc.loc.exists()) {
+                        /**
+                         * Sends the waypoints message except it suggests which one should be used based on
+                         * the name contained in the message and converts it to the internally used names for the waypoints.
+                         */
+                        UMessage("§3Skytils > §eFound coordinates in a chat message, click a button to set a waypoint.\n")
+                            .append(UTextComponent("§f${loc.displayName} ")
+                                .setClick(MCClickEventAction.RUN_COMMAND, "/skytilshollowwaypoint set ${loc.id} $x $y $z")
+                                .setHoverText("§eSet waypoint for ${loc.displayName}"))
+                            .append(UTextComponent("§e[Custom]")
+                                .setClick(MCClickEventAction.SUGGEST_COMMAND, "/skytilshollowwaypoint set name_here $x $y $z")
+                                .setHoverText("§eSet custom waypoint")
+                            ).chat()
+                    }
```

Comment:
you can rewrite this into just a let
```suggestion
                CrystalHollowsMap.Locations.values().find { it.cleanName == stringLocation }?.takeIf { !it.loc.exists() }?.let{ loc ->
                    /**
                     * Sends the waypoints message except it suggests which one should be used based on
                     * the name contained in the message and converts it to the internally used names for the waypoints.
                     */
                    UMessage("§3Skytils > §eFound coordinates in a chat message, click a button to set a waypoint.\n")
                        .append(UTextComponent("§f${loc.displayName} ")
                            .setClick(MCClickEventAction.RUN_COMMAND, "/skytilshollowwaypoint set ${loc.id} $x $y $z")
                            .setHoverText("§eSet waypoint for ${loc.displayName}"))
                        .append(UTextComponent("§e[Custom]")
                            .setClick(MCClickEventAction.SUGGEST_COMMAND, "/skytilshollowwaypoint set name_here $x $y $z")
                            .setHoverText("§eSet custom waypoint")
                        ).chat()
```

Reply:
sry again im new to kotlin... didn't know these functions were a thing